File: src/main/java/com/minecraft/moonlake/nbt/NBTBase.java
/*
 * Copyright (C) 2017 The MoonLake Authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


package com.minecraft.moonlake.nbt;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;

/**
 * <h1>NBTBase</h1>
 * NBT 标签
 *
 * @version 1.0
 * @author Month_Light
 */
public abstract class NBTBase implements Cloneable {

    private final String name;

    /**
     * NBT 标签构造函数
     *
     * @param name 特殊名
     */
    public NBTBase(String name) {
        this.name = name;
    }

    /**
     * 获取此 NBT 标签特殊名
     *
     * @return 特殊名
     */
    public final String getName() {
        return name;
    }

    /**
     * 获取此 NBT 标签的值
     *
     * @return 值
     */
    public abstract Object getValue();

    /**
     * 获取此 NBT 标签类型
     *
     * @return 类型
     */
    public abstract NBTType getType();

    /**
     * 读取指定数据输入流
     *
     * @param input 数据输入流
     * @throws IOException 如果 IO 错误则抛出异常
     */
    public abstract void read(DataInput input) throws IOException;

    /**
     * 写入指定数据输出流
     *
     * @param output 数据输出流
     * @throws IOException 如果 IO 错误则抛出异常
     */
    public abstract void write(DataOutput output) throws IOException;

    /**
     * 获取此 NBT 标签类型 Id
     *
     * @return 类型 Id
     */
    public int getTypeId() {
        return getType().getTypeId();
    }

    @Override
    public boolean equals(Object obj) {
        if(obj == this)
            return true;
        if(obj instanceof NBTBase) {
            NBTBase other = (NBTBase) obj;
            return name.equals(other.name) && getTypeId() == other.getTypeId();
        }
        return false;
    }

    @Override
    public int hashCode() {
        return getTypeId();
    }

    @Override
    public abstract String toString();

    @Override
    public abstract NBTBase clone();
}

File: API/src/main/kotlin/com/mcmoonlake/api/nbt/NBTBase.kt
/*
 *  ProtocolLib - Bukkit server library that allows access to the Minecraft protocol.
 *  Copyright (C) 2012 Kristian S. Stangeland
 *
 *  This program is free software; you can redistribute it and/or modify it under the terms of the
 *  GNU General Public License as published by the Free Software Foundation; either version 2 of
 *  the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *  See the GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along with this program;
 *  if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 *  02111-1307 USA
 */

/*
 * Copyright (C) 2016-Present The MoonLake (mcmoonlake@hotmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.mcmoonlake.api.nbt

interface NBTBase<T> {

    var name: String

    val type: NBTType

    var value: T

    // TODO v1.13 NBT Color Enhancement & Deprecated Remove ?
    // Color Enhancement Completed
    /**
     * Convert the json string to the numeric carrying type suffix.
     *
     * @param color Whether it has a color of 1.13
     */
    @Deprecated("The future version 1.13 may be subject to change.") // TODO v1.13
    fun toMojangson(color: Boolean = false): String

    /**
     * Convert the json string to the numeric carrying type suffix.
     */
    @Deprecated("The future version 1.13 may be subject to change.") // TODO v1.13
    fun toMojangson(): String
            = toMojangson(false)
}
